fs: fix rmsync error swallowing

Linkgoron opened the pull request:
Fix `rmsync` swallowing errors instead of throwing them. Note that I've also changed a few places that retry errors, to align `rmsync` to `rm`.

If anyone has a better idea for a test, I'd be happy to implement it - as changing permissions feels like it might cause errors if the file somehow survives.

fixes: #38683
fixes: #34580

VoltrexMaster commented on test/parallel/test-fs-rm.js:
```diff
@@ -280,3 +280,24 @@ function removeAsync(dir) {
 }
+
+{
+  // Check the deleting a file that cannot be accessed using rmsync throws
```
```suggestion
  // Check if deleting a file that cannot be accessed using rmsync throws
```

aduh95 commented on lib/internal/fs/rimraf.js:
```diff
@@ -219,6 +219,12 @@ function _unlinkSync(path, options) {
           i < tries &&
           options.retryDelay > 0) {
         sleep(i * options.retryDelay);
+      } else {
+        // The file is already gone.
+        if (err.code === 'ENOENT') {
+          return;
+        }
+        throw err;
```
nit: we could avoid nested `if` here.
```suggestion
      } else if (err.code !== 'ENOENT') {
        throw err;
      } else {
        // The file is already gone.
        return;
```

aduh95 commented on lib/internal/fs/rimraf.js:
```diff
@@ -259,10 +265,18 @@ function _rmdirSync(path, options, originalErr) {
               i < tries &&
               options.retryDelay > 0) {
             sleep(i * options.retryDelay);
+          } else {
+            // The file is already gone.
+            if (err.code === 'ENOENT') {
+              return;
+            }
+            throw err;
```
nit: we could avoid nested `if` here.
```suggestion
      } else if (err.code !== 'ENOENT') {
        throw err;
      } else {
        // The file is already gone.
        return;
```

RaisinTen commented on test/parallel/test-fs-rm.js:
```diff
@@ -280,3 +280,24 @@ function removeAsync(dir) {
+      code: 'EACCES',
+      syscall: 'unlink',
+      name: 'Error',
+      errno: -13,
```
```suggestion
```
`errno` can hold different values on different platforms, so I think we should just stick to `code`.

RaisinTen commented:
I had a feeling the Windows tests would fail. Not quite sure why they are failing though. I had a PR open that has a similar test using the `chmod + rm` combo in order to fix another `rm` bug and even that fails: https://github.com/nodejs/node/pull/36815.

Linkgoron commented:
> I had a feeling the Windows tests would fail. Not quite sure why they are failing though. I had a PR open that has a similar test using the `chmod + rm` combo in order to fix another `rm` bug and even that fails: #36815.

Yep, and it failed to remove the file which caused other test to fail as well

RaisinTen commented on test/parallel/test-fs-rm.js:
```diff
@@ -280,3 +282,44 @@ function removeAsync(dir) {
 }
+
+{
+  if (!common.isIBMi && (common.isWindows || process.getuid() !== 0)) {
```
I'm guessing this is there to stop this test from running on IBMi and Android. Could you please explain why we are doing this?

Linkgoron replied:
That's what other 'EPERM' and 'EACCESS' tests do

RaisinTen replied:
Ah, to prevent it from being run with root privileges. Yes, I think it would be more clear if we added in that comment.

RaisinTen commented on test/parallel/test-fs-rm.js:
```diff
@@ -280,3 +282,44 @@ function removeAsync(dir) {
+      let accessErrorCode = 'EACCES';
+      if (common.isWindows) {
+        accessErrorCode = 'EPERM';
+        execSync(`icacls ${dir} /inheritance:r`);
+        execSync(`icacls ${dir} /deny "everyone":W`);
```
```suggestion
        execSync(`icacls ${dir} /deny "everyone:(OI)(CI)(DE,DC)"`);
```
Maybe we need to explicitly set the delete permissions?

RaisinTen commented on test/parallel/test-fs-rm.js:
```diff
@@ -280,3 +282,44 @@ function removeAsync(dir) {
+
+    function makeDirectoryWritable(dir) {
+      if (common.isWindows) {
+        execSync(`icacls ${dir} /grant "everyone":W`);
```
```suggestion
        execSync(`icacls ${dir} /remove:d "everyone"`);
```

Linkgoron replied:
This worked locally, so hopefully this will also work in CI

aduh95 commented:
Windows CI's still failing.

Linkgoron commented:
> Windows CI's still failing.

ARM is also consistantly failing

Linkgoron commented:
@Trott do you think that there's any chance that this fix uncovered some things that didn't really work on Windows? Some calls to `rmSync` that were errored but didn't throw?

Trott commented:
> @Trott do you think that there's any chance that this fix uncovered some things that didn't really work on Windows? Some calls to `rmSync` that were errored but didn't throw?

¯\\_(ツ)_/¯

@cjihrig

cjihrig commented:
Sorry, but I'm also ¯\\(ツ)/¯ here, especially on Windows.

Linkgoron commented:
Looks like this failed for every test that didn't close all of its open fds. I _think_ that I've caught all of them.

Linkgoron commented:
Given all of the issues encountered, and assuming they were actually fixed, I think that it makes sense to mark this as semver-major.

Trott commented:
@nodejs/platform-windows

Linkgoron commented:
Tests on Windows now pass, https://ci.nodejs.org/job/node-test-binary-windows-js-suites/9966/

aduh95 commented:
> Given all of the issues encountered, and assuming they were actually fixed, I think that it makes sense to mark this as semver-major.

@Linkgoron do you have an example of code that works in a current Node.js version and would throw with this PR? Is this something we should document?

Linkgoron commented:
> > Given all of the issues encountered, and assuming they were actually fixed, I think that it makes sense to mark this as semver-major.
> 
> @Linkgoron do you have an example of code that works in a current Node.js version and would throw with this PR? Is this something we should document?

The test that I added in ` test/parallel/test-fs-rm.js` doesn't throw in current Node versions, even though it should throw (rm already gives an error in the callback). As we've seen with the test fixes, It appears that trying to delete a folder with an open file also throws on Windows and maybe ARM (doesn't throw on Mac or Linux).

RaisinTen commented on test/parallel/test-fs-mkdir-recursive-eaccess.js:
```diff
@@ -26,8 +26,7 @@ function makeDirectoryReadOnly(dir) {
     accessErrorCode = 'EPERM';
-    execSync(`icacls ${dir} /inheritance:r`);
-    execSync(`icacls ${dir} /deny "everyone":W`);
+    execSync(`icacls ${dir} /deny "everyone:(OI)(CI)(DE,DC,AD,WD)"`);
```
```suggestion
    execSync(`attrib +r ${dir}`);
```

Linkgoron replied:
This doesn't work locally for me (doesn't throw).

RaisinTen replied:
Does it also remove the file successfully?

Linkgoron replied:
Yes, I also tried with /s and /d

Linkgoron replied:
It's fine, I don't mind experimenting and simplifying things to see if they work out.

RaisinTen replied:
Also, `icacls` doesn't stop me from deleting the directory: https://github.com/RaisinTen/tests/runs/2602063819?check_suite_focus=true

Linkgoron replied:
Specifically, the test doesn't try  to remove the folder - one test tries to create a subfolder and one tries to remove a file inside of the directory.

Linkgoron replied:
I think that DE might be wrong, as it's not listed here:
https://docs.microsoft.com/en-us/windows-server/administration/windows-commands/icacls
It _is_ listed in older docs that I found.

RaisinTen commented on test/parallel/test-fs-rm.js:
```diff
@@ -280,3 +282,45 @@ function removeAsync(dir) {
+      let accessErrorCode = 'EACCES';
+      if (common.isWindows) {
+        accessErrorCode = 'EPERM';
+        execSync(`icacls ${dir} /deny "everyone:(OI)(CI)(DE,DC)"`);
```
```suggestion
        execSync(`attib +r ${dir}`);
```

RaisinTen commented on test/parallel/test-fs-rm.js:
```diff
@@ -280,3 +282,45 @@ function removeAsync(dir) {
+
+    function makeDirectoryWritable(dir) {
+      if (common.isWindows) {
+        execSync(`icacls ${dir} /remove:d "everyone"`);
```
```suggestion
        execSync(`attrib -r ${dir}`);
```

RaisinTen commented:
Libuv's implementation of `chmod` uses [`NtQueryInformationFile`](https://docs.microsoft.com/en-us/windows-hardware/drivers/ddi/ntifs/nf-ntifs-ntqueryinformationfile) with the `FileInformationClass` set to [`FileBasicInformation`](https://docs.microsoft.com/en-us/windows-hardware/drivers/ddi/wdm/ns-wdm-_file_basic_information) to retrieve the `FileInformation`. Then it updates `FileInformation` with the [`FILE_ATTRIBUTE_READONLY`](https://docs.microsoft.com/en-us/windows/win32/fileio/file-attribute-constants) mask. Finally, it updates the `FileInformation` of the file using [`NtSetInformationFile`](https://docs.microsoft.com/en-us/windows-hardware/drivers/ddi/ntifs/nf-ntifs-ntsetinformationfile) with the `FileInformationClass` set to [`FileBasicInformation`](https://docs.microsoft.com/en-us/windows-hardware/drivers/ddi/wdm/ns-wdm-_file_basic_information).

So, I think, Libuv's `chmod` works in a way more similar to `attrib` than the way `icacls` works. That's why I think we should check whether this works.

RaisinTen commented on test/parallel/test-fs-rm.js:
```diff
@@ -280,3 +282,45 @@ function removeAsync(dir) {
+      if (common.isWindows) {
+        execSync(`icacls ${dir} /remove:d "everyone"`);
+      } else {
+        fs.chmodSync(dirname, 0o777);
```
```suggestion
        fs.chmodSync(dir, 0o777);
```
typo?

Linkgoron replied:
yep.

RaisinTen commented:
> Libuv's implementation of `chmod` uses [`NtQueryInformationFile`](https://docs.microsoft.com/en-us/windows-hardware/drivers/ddi/ntifs/nf-ntifs-ntqueryinformationfile) with the `FileInformationClass` set to [`FileBasicInformation`](https://docs.microsoft.com/en-us/windows-hardware/drivers/ddi/wdm/ns-wdm-_file_basic_information) to retrieve the `FileInformation`. Then it updates `FileInformation` with the [`FILE_ATTRIBUTE_READONLY`](https://docs.microsoft.com/en-us/windows/win32/fileio/file-attribute-constants) mask. Finally, it updates the `FileInformation` of the file using [`NtSetInformationFile`](https://docs.microsoft.com/en-us/windows-hardware/drivers/ddi/ntifs/nf-ntifs-ntsetinformationfile) with the `FileInformationClass` set to [`FileBasicInformation`](https://docs.microsoft.com/en-us/windows-hardware/drivers/ddi/wdm/ns-wdm-_file_basic_information).
> 
> So, I think, Libuv's `chmod` works in a way more similar to `attrib` than the way `icacls` works. That's why I think we should check whether this works.

There's already an open issue in Libuv (https://github.com/libuv/libuv/issues/2838) to use ACL in `chmod`. So, I guess it is better to stick to `icacls` till there is a change from Libuv's side.

RaisinTen commented on test/parallel/test-fs-mkdir-recursive-eaccess.js:
```diff
@@ -26,8 +26,7 @@ function makeDirectoryReadOnly(dir) {
     accessErrorCode = 'EPERM';
-    execSync(`icacls ${dir} /inheritance:r`);
-    execSync(`icacls ${dir} /deny "everyone":W`);
+    execSync(`icacls ${dir} /deny "everyone:(OI)(CI)(DE,DC,AD,WD)"`);
```
```suggestion
    execSync(`icacls ${dir} /deny "everyone:(OI)(CI)(D,DC,AD,WD)"`);
```

RaisinTen commented on test/parallel/test-fs-rm.js:
```diff
@@ -280,3 +282,45 @@ function removeAsync(dir) {
+      let accessErrorCode = 'EACCES';
+      if (common.isWindows) {
+        accessErrorCode = 'EPERM';
+        execSync(`icacls ${dir} /deny "everyone:(OI)(CI)(DE,DC)"`);
```
```suggestion
        execSync(`icacls ${dir} /deny "everyone:(OI)(CI)(D,DC)"`);
```

Linkgoron replied:
Apparently `D` breaks the test, so I changed it back to DE

RaisinTen commented:
I think this PR also fixes https://github.com/nodejs/node/issues/34580. Could you please add a test to verify it?

Linkgoron commented:
> #34580

It needs an additional fix, but I'll add it

aduh95 commented on lib/internal/fs/rimraf.js:
```diff
@@ -231,8 +236,12 @@ function _rmdirSync(path, options, originalErr) {
-    if (err.code === 'ENOTDIR')
-      throw originalErr;
+    if (err.code === 'ENOTDIR') {
+      if (originalErr) {
+        throw originalErr;
+      }
+      throw err;
```
```suggestion
      throw originalErr || err;
```

aduh95 commented on lib/internal/fs/rimraf.js:
```diff
@@ -259,10 +268,21 @@ function _rmdirSync(path, options, originalErr) {
       }
     }
+
+    if (originalErr) {
+      throw originalErr;
+    } else {
+      throw err;
+    }
```
```suggestion
    throw originalErr || err;
```

aduh95 reviewed (approved):
nit

Linkgoron commented:
As an aside, I don't get why rimraf tries to fix EPERM on windows by calling `chmod(0o666)` and yet we don't do the same on EACCES on POSIX.

RaisinTen commented:
> As an aside, I don't get why rimraf tries to fix EPERM on windows by calling `chmod(0o666)`

Because isaacs was convinced: https://github.com/isaacs/rimraf/issues/21#issuecomment-29526406

I think the main reason is that `.git` was (is? idk) read-only on Windows: https://github.com/isaacs/rimraf/issues/19

PR that introduced this: https://github.com/isaacs/rimraf/pull/26

> yet we don't do the same on EACCES on POSIX (actually, what I don't get is why it doesn't throw in both cases).

I don't think there was ever any `EACCES` issue with the `.git` directory on POSIX.

RaisinTen commented:
Still fails on Windows, removing the `author-ready`. :/

Linkgoron commented:
> Still fails on Windows, removing the `author-ready`. :/

It only fails on one specific machine, and doesn't reproduce locally for me. Still trying to figure out what's wrong. At first I thought that it successfully removes the files, so I changed the test - but it appears that it's not the case.

RaisinTen commented:
@linkgoron What do you think about reverting the changes you did after https://github.com/nodejs/node/pull/38684#issuecomment-844114260? After learning about the Windows specific EPERM fix, I was thinking about removing that (in a separate PR). That way, `rm` would not change any file permissions and would work the same way across all platforms and would fix the busy wait problem too.

Linkgoron commented:
> @Linkgoron What do you think about reverting the changes you did after [#38684 (comment)](https://github.com/nodejs/node/pull/38684#issuecomment-844114260)? After learning about the Windows specific EPERM fix, I was thinking about removing that (in a separate PR). That way, `rm` would not change any file permissions and would work the same way across all platforms and would fix the busy wait problem too.

What I fixed was a real issue in my original code, all I changed was throwing `err` if `originalErr` is null.

RaisinTen commented:
> > Still fails on Windows, removing the `author-ready`. :/
> 
> It only fails on one specific machine, and doesn't reproduce locally for me. Still trying to figure out what's wrong. At first I thought that it successfully removes the files, so I changed the test - but it appears that it's not the case.

Probably on that machine, `attrib` (from the `chmod` call in `fixWinEPERM`) can revert the changes made by `icacls` in `makeDirectoryReadOnly`.

Linkgoron commented:
> > > Still fails on Windows, removing the `author-ready`. :/
> > 
> > 
> > It only fails on one specific machine, and doesn't reproduce locally for me. Still trying to figure out what's wrong. At first I thought that it successfully removes the files, so I changed the test - but it appears that it's not the case.
> 
> Probably on that machine, `attrib` (from the `chmod` call in `fixWinEPERM`) can revert the changes made by `icacls` in `makeDirectoryReadOnly`.

If that were the case, I think that my modified test should have passed.

RaisinTen commented on test/parallel/test-fs-rm.js:
```diff
@@ -280,3 +282,80 @@ function removeAsync(dir) {
+            name: 'Error',
+          });
+        } catch (err) {
+          // Only fail the test if the folder was not deleted.
+          // as in some cases rmSync succesfully deletes read-only folders.
+          if (!fs.existsSync(root)) {
+            throw err;
+          }
```
As per the comment, this should `throw err` if:
`root` was not deleted = `root` still exists
But as per the conditional, `if (!fs.existsSync(root)) {`:
This branch is executed if `root` does not exist.
So, shouldn't this be:
```suggestion
          // Only fail the test if the folder was not deleted.
          // as in some cases rmSync succesfully deletes read-only folders.
          if (fs.existsSync(root)) {
            throw err;
          }
```

Linkgoron replied:
whoops :P

Linkgoron replied:
-_-

richardlau commented:
Is it possible this has caused the `parallel/test-fs-writefile-with-fd` Windows failures in today's daily master CI and other recent CIs? 
https://ci.nodejs.org/job/node-test-binary-windows-js-suites/10075/#showFailuresLink
```
Can't clean tmpdir: C:\workspace\node-test-binary-windows-js-suites\node\test\.tmp.262
Files blocking: [ 'test.txt' ]

C:\workspace\node-test-binary-windows-js-suites\node\test\common\tmpdir.js:55
    throw e;
    ^

Error: ENOTEMPTY: directory not empty, rmdir '\\?\C:\workspace\node-test-binary-windows-js-suites\node\test\.tmp.262'
    at Object.rmdirSync (node:fs:1158:10)
    at _rmdirSync (node:internal/fs/rimraf:260:21)
    at rimrafSync (node:internal/fs/rimraf:193:7)
    at Object.rmSync (node:fs:1205:10)
    at rmSync (C:\workspace\node-test-binary-windows-js-suites\node\test\common\tmpdir.js:9:6)
    at process.onexit (C:\workspace\node-test-binary-windows-js-suites\node\test\common\tmpdir.js:40:5)
    at process.emit (node:events:406:35) {
  errno: -4051,
  syscall: 'rmdir',
  code: 'ENOTEMPTY',
  path: '\\\\?\\C:\\workspace\\node-test-binary-windows-js-suites\\node\\test\\.tmp.262'
}
```

targos commented:
I also suspect this change.

Linkgoron commented:
yes, it's 100% this change - the test opens test.txt but doesn't close the `fd` which causes errors on Windows when trying to remove the directory. I'm not sure how it passed the other tests.

richardlau commented:
Yeah, arm-fanned failed the same `parallel/test-fs-writefile-with-fd` test as it runs on an NFS-mounted FS: https://ci.nodejs.org/job/node-test-binary-arm-12+/11011/
```
Can't clean tmpdir: /home/iojs/tmp/.tmp.261
Files blocking: [ '.nfs00000000013426cf00004f89' ]
Note: ".nfs*" might be files that were open and unlinked but not closed.
See http://nfs.sourceforge.net/#faq_d2 for details.

/home/iojs/build/workspace/node-test-binary-arm/test/common/tmpdir.js:55
    throw e;
    ^

Error: ENOTEMPTY: directory not empty, rmdir '/home/iojs/tmp/.tmp.261'
    at Object.rmdirSync (node:fs:1158:10)
    at _rmdirSync (node:internal/fs/rimraf:260:21)
    at rimrafSync (node:internal/fs/rimraf:193:7)
    at Object.rmSync (node:fs:1205:10)
    at rmSync (/home/iojs/build/workspace/node-test-binary-arm/test/common/tmpdir.js:9:6)
    at process.onexit (/home/iojs/build/workspace/node-test-binary-arm/test/common/tmpdir.js:40:5)
    at process.emit (node:events:406:35) {
  errno: -39,
  syscall: 'rmdir',
  code: 'ENOTEMPTY',
  path: '/home/iojs/tmp/.tmp.261'
}
```

Linkgoron commented:
fix here: https://github.com/nodejs/node/pull/38820
